getretake/retake
# feat: Introduce realtime data sync

**Ticket(s) Closed**

- Closes #94 

**What**
This PR adds realtime sync  between postgres and the search api. It uses kafka and debezium to pipe changes as they happen on the database and the api takes care of processing and inserting to the index.

Also caught and fixed the `model not deployed` error - we now deploy the model and wait for it in `register_neural_search_fields`

**Why**

**How**

Testing instructions:
1. Start the stack using `docker compose up --build`
2. Start a postgres container with `docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=postgres -d postgres -c wal_level=logical` or use an existing postgres database.
3. If using the container, connect to the compose network with `docker networks connect search_default postgres`
4. Call the realtime link endpoint:
```bash
curl --location 'http://localhost:8000/client/realtime/link' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <api_key>' \
--data '{
    "source_host": "postgres",
    "source_port": "5432",
    "source_user": "postgres",
    "source_password": "postgres",
    "source_relation": <relation>,
    "source_primary_key": <primary_key>,
    "source_columns": ["col1", "col2"],
    "source_neural_columns": ["col1"]
}'
```
5. It will take some time to process. After it has finished, call the realtime start endpoint to start processing events in real time:

```bash
curl --location 'http://localhost:8000/client/realtime/start' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <api_key>' \
--data '{
    "source_host": "postgres",
    "source_port": "5432",
    "source_user": "postgres",
    "source_password": "postgres",
    "source_relation": <relation>,
    "source_primary_key": <primary_key>,
    "source_columns": ["col1", "col2"],
    "source_neural_columns": ["col1"]
}'
```
6. You should get a successful response immediately, and the processing will happen in the background. Check the api logs with `docker compose logs api`
7. Confirm the records were stored in the index successfully:
- Run `docker cp <opensearch_container_id>:/usr/share/opensearch/config/root-ca.pem .`
- `curl -XPOST -H "Content-Type: application/json" -u admin:admin --cacert root-ca.pem https://localhost:9200/<index_name>/_search`

**Tests**
